### ISSUE TITLE
Re-enable chained proxy checks

### ISSUE THREAD

**Issue by myleshorton** [getlantern/lantern#4268]
Checks are disabled [here](https://github.com/getlantern/lantern/blob/devel/src/github.com/getlantern/balancer/dialer.go#L78) but should be enabled certainly for pro.

They were disabled because we were effectively DDOSing servers and were blacklisting clients. Some limited discussion at #4261

**Comment by oxtoacart** [comment]
Consider other options like maybe randomly sending requests through lower priority servers.

**Comment by fffw** [comment]
So the basic idea is:

- [ ] 1. Not checking against Google.

      1. Check against the `pinger` in chained dialer.
      2. Remember the last address it dialed if the port is 80 or 443. Have the check code to issue a GET request to `index.html` and see if it returns non 5XX codes.

- [ ] 2. Reproduce https://github.com/getlantern/lantern/issues/4261 and fix accordingly.

**Comment by fffw** [comment]
It's non-trival to tunnel HTTPS at this level, but plain HTTP requests are very few, most of them are requests to `config.getiantem.org`. Maybe we can test against `pinger` most of the time.

**Comment by fffw** [comment]
Regarding why the checker opens a connection for 30s without sending anything, there are various steps between opening a connection and sending HTTP request. Listed as below:

> [An `http.Client`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/balancer/dialer.go#L149) with an `http.Transport`, with it's `Dial` set as

> > [`balancer.dialer.dial`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/balancer/dialer.go#L152), which calls

> > > [`balancer.dialer.DialFN`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/balancer/dialer.go#L108). The dialer is returned from

> > > > [`chained.NewDialer`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/flashlight/client/chained.go#L87), which does 2 things:

> > > > > call [`DialServer`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/chained/dialer.go#L41), which is returned by

> > > > > > [`dialFactory`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/flashlight/client/chained.go#L59), which is one of below depends on the config:

> > > > > > > [`defaultDialFactory`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/flashlight/client/pluggable_transports.go#L23)

> > > > > > > [`obfs4DialFactory`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/flashlight/client/pluggable_transports.go#L23)

> > > > > ~~[`sendCONNECT`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/chained/dialer.go#L50) if the network is a special string `connect`, which doesn't apply here.~~

> > > Wrap to [`idletiming.Conn`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/flashlight/client/chained.go#L98)

> [`client.Do`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/balancer/dialer.go#L162)

The whole procedure is wrapped in [`withtimeout.Do`](https://github.com/getlantern/lantern/blob/2.2.1/src/github.com/getlantern/balancer/dialer.go#L155) with a 1-minute timeout. 

So the question becomes: What happened during the time between calling `DialServer` and `client.Do`?

**Comment by fffw** [comment]
A normal proxy request dials server via `balancer.Balancer.Dial` which also goes through `balancer.dialer.dial`, but we didn't see the same phenomenon for normal requests. `withtimeout.Do` should have nothing to do with the issue.

**Comment by fffw** [comment]
I was trying to log the time required to finish TLS handshake, and the `tls.ConnectionState`, in hope to check if it's the TLS handshake took a long time. However, handshake is not part of the logic in `tls.Listener.Accept()`, so there's no way to measure it from application code.

Actually, `tlsConn.Handshake` is called in [`conn.serve()`](https://github.com/golang/go/blob/release-branch.go1.6/src/net/http/server.go#L1405), which is even between `StateNew` and `StateActive` when calling the supplied `http.Server.ConnState` callback.

Dead end, need other way around. From my experience, TLS handshake can take a long time sometimes, so it still could be the reason.

**Comment by oxtoacart** [comment]
@fffw If you want, you could temporarily use your own custom listener that manually handshakes inside of its Accept implementation so that you can record the timing.

**Comment by oxtoacart** [comment]
@fffw This is not a permanent solution because it slows down the server by blocking on the TLS handshake, meaning that subsequent connections won't be accepted until the current handshake finishes.

**Comment by fffw** [comment]
Thanks @oxtoacart . `http.Server` does handshake in `conn.serve()`. Manual handshake may interfere with it but I'm not sure. Anyway, I changed my mind to do something at client side instead.

**Comment by oxtoacart** [comment]
@fffw Okay.  FWIW, tls.Conn is smart enough to handshake only once, so once you've called handshake() the subsequent call to handshake does nothing.

**Comment by fffw** [comment]
Wow I never looked into Go code that deeply. Good to know it. Thanks.
